internal/ethapi/api.go

```diff
@@ -1648,6 +1753,161 @@ func SubmitTransaction(ctx context.Context, b Backend, tx *types.Transaction) (c
 		log.EmitCheckpoint(log.TxCreated, "tx", tx.Hash().Hex(), "to", tx.To().Hex())
 	}
 	return tx.Hash(), nil
+
+}
+
+// Quorum
+//
+// performMultitenancyChecks is to use the given transaction and construct
+// expected security attributes being checked against entitled ones. The
+// transaction is fed into the simulation engine in order to determine the impact.
+func performMultitenancyChecks(ctx context.Context, authToken *proto.PreAuthenticatedAuthenticationToken,
+	b Backend, fromEOA common.Address, tx *types.Transaction, privateArgs *PrivateTxArgs) error {
+
+	if tx.IsPrivate() {
+		// before running simulation, we verify the ownership of privateFrom
+		// user must be entitled for all actions
+		// READ and WRITE actions are taking Parties into consideration so
+		// we need to populate it with privateFrom
+		if authorized, _ := b.IsAuthorized(ctx, authToken,
+			multitenancy.NewContractSecurityAttributeBuilder().FromEOA(fromEOA).Private().Create().PrivateFrom(privateArgs.PrivateFrom).Build(),
+			multitenancy.NewContractSecurityAttributeBuilder().FromEOA(fromEOA).Private().Write().Party(privateArgs.PrivateFrom).Build(),
+			multitenancy.NewContractSecurityAttributeBuilder().FromEOA(fromEOA).Private().Read().Party(privateArgs.PrivateFrom).Build(),
+		); !authorized {
+			return multitenancy.ErrNotAuthorized
+		}
+	}
+	currentBlock := b.CurrentBlock().Number().Int64()
+	extraDataReader, err := b.AccountExtraDataStateReaderByNumber(ctx, rpc.BlockNumber(currentBlock))
+	if err != nil {
+		return fmt.Errorf("no account extra data reader at block %v: %w", currentBlock, err)
+	}
+	// create callbacks to support runtime multitenancy checks during simulation
+	createContractSA := multitenancy.NewContractSecurityAttributeBuilder().
+		FromEOA(fromEOA).PrivateIf(tx.IsPrivate()).Create().PrivateFromOnlyIf(tx.IsPrivate(), privateArgs.PrivateFrom).Build()
+	authorizedCreate, _ := b.IsAuthorized(ctx, authToken, createContractSA)
+	var authorizeCreateFunc multitenancy.AuthorizeCreateFunc = func() bool {
+		return authorizedCreate
+	}
+	var authorizeMessageCallFunc multitenancy.AuthorizeMessageCallFunc = func(contractAddress common.Address) (bool, bool, error) {
+		managedParties, err := extraDataReader.ReadManagedParties(contractAddress)
+		isPrivate := true
+		if errors.Is(err, common.ErrNotPrivateContract) {
+			isPrivate = false
+		} else if err != nil {
+			return false, false, fmt.Errorf("%s not found in the index, error: %s", contractAddress.Hex(), err.Error())
+		}
+		readSecAttr := multitenancy.NewContractSecurityAttributeBuilder().FromEOA(fromEOA).PrivateIf(isPrivate).PartiesOnlyIf(isPrivate, managedParties).Read().Build()
```

Comment:
MINOR: I don't think we need to declare `readSecAttr` could be merge in `IsAuthorized` call